x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheServiceTests.java

```diff
@@ -221,44 +226,150 @@ public void testPut() throws Exception {
         if (randomBoolean()) {
-            cacheService.markShardAsEvictedInCache(snapshotId, indexId, shardId); // no effect
+            // mark shard as evicted after cache service is stopped should have no effect
+            cacheService.markShardAsEvictedInCache(shard.getSnapshotUUID(), shard.getSnapshotIndexName(), shard.getShardId());
+            assertThat(cacheService.pendingShardsEvictions(), aMapWithSize(0));
         }
-        waitForEviction.get(30L, TimeUnit.SECONDS);
-        cacheFile.release(evictionListener);
-
-        cacheService.runIfShardMarkedAsEvictedInCache(
-            snapshotId,
-            indexId,
-            shardId,
-            () -> { assert false : "should not be called: shard eviction marker is removed"; }
+    }
+
+    public void testProcessShardEviction() throws Exception {
+        final CacheService cacheService = defaultCacheService();
+        cacheService.start();
+
+        final List<CacheFile> randomCacheFiles = randomCacheFiles(cacheService);
+        assertThat(cacheService.pendingShardsEvictions(), aMapWithSize(0));
+
+        final ShardEviction shard = randomShardEvictionFrom(randomCacheFiles);
+        final List<CacheFile> cacheFilesAssociatedWithShard = filterByShard(shard, randomCacheFiles);
+        cacheFilesAssociatedWithShard.forEach(cacheFile -> assertTrue(Files.exists(cacheFile.getFile())));
+
+        final BlockingEvictionListener blockingListener = new BlockingEvictionListener();
+        final CacheFile randomCacheFile = randomFrom(cacheFilesAssociatedWithShard);
+        assertTrue(Files.exists(randomCacheFile.getFile()));
+        randomCacheFile.acquire(blockingListener);
+
+        cacheService.markShardAsEvictedInCache(shard.getSnapshotUUID(), shard.getSnapshotIndexName(), shard.getShardId());
+
+        final Map<CacheFile, Boolean> afterShardRecoveryCacheFiles = ConcurrentCollections.newConcurrentMap();
+        final Thread shardRecoveryThread = new Thread(() -> {
+            cacheService.waitForCacheFilesEvictionIfNeeded(shard.getSnapshotUUID(), shard.getSnapshotIndexName(), shard.getShardId());
+            for (CacheFile cacheFile : cacheFilesAssociatedWithShard) {
+                afterShardRecoveryCacheFiles.put(cacheFile, Files.exists(cacheFile.getFile()));
+            }
+        });
+        shardRecoveryThread.start();
+
+        blockingListener.waitForBlock();
+
+        final Map<ShardEviction, Future<?>> pendingShardsEvictions = cacheService.pendingShardsEvictions();
+        assertTrue(cacheService.isPendingShardEviction(shard));
+        assertThat(pendingShardsEvictions, aMapWithSize(1));
+
+        final Future<?> shardEvictionFuture = pendingShardsEvictions.get(shard);
+        assertTrue(Files.exists(randomCacheFile.getFile()));
+        assertThat(shardEvictionFuture, notNullValue());
+        assertFalse(shardEvictionFuture.isDone());
+
+        blockingListener.unblock();
+
+        FutureUtils.get(shardEvictionFuture);
```

Comment:
Maybe we can move this below the join and then first:

```
assertTrue(shardEvictionFuture.isDone());
```

Reply:
I had to adapt this part to play with the `generic` thread pool assertion but I moved this below what was the join (now a get on the future returned by submitting a runnable to the generic tp) -> b9cfded3a8f4d4f3877c9de7d15c5e54c464ce2d